File: backend/app/models/capture.py
"""
RFP Sniper - Capture Models
===========================
Capture pipeline, bid decisions, and teaming partners.
"""

from datetime import date, datetime
from enum import Enum

from sqlmodel import JSON, Column, Field, SQLModel, Text


class CaptureStage(str, Enum):
    IDENTIFIED = "identified"
    QUALIFIED = "qualified"
    PURSUIT = "pursuit"
    PROPOSAL = "proposal"
    SUBMITTED = "submitted"
    WON = "won"
    LOST = "lost"


class BidDecision(str, Enum):
    PENDING = "pending"
    BID = "bid"
    NO_BID = "no_bid"


class CapturePlan(SQLModel, table=True):
    """
    Capture plan for an opportunity.
    """

    __tablename__ = "capture_plans"

    id: int | None = Field(default=None, primary_key=True)
    rfp_id: int = Field(foreign_key="rfps.id", index=True, unique=True)
    owner_id: int = Field(foreign_key="users.id", index=True)

    stage: CaptureStage = Field(default=CaptureStage.IDENTIFIED)
    bid_decision: BidDecision = Field(default=BidDecision.PENDING)
    win_probability: int | None = Field(default=None, ge=0, le=100)
    notes: str | None = None

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class GateReview(SQLModel, table=True):
    """
    Capture gate review record.
    """

    __tablename__ = "gate_reviews"

    id: int | None = Field(default=None, primary_key=True)
    rfp_id: int = Field(foreign_key="rfps.id", index=True)
    reviewer_id: int = Field(foreign_key="users.id", index=True)

    stage: CaptureStage = Field(default=CaptureStage.QUALIFIED)
    decision: BidDecision = Field(default=BidDecision.PENDING)
    notes: str | None = None

    created_at: datetime = Field(default_factory=datetime.utcnow)


class TeamingPartner(SQLModel, table=True):
    """
    Teaming partner directory.
    """

    __tablename__ = "teaming_partners"

    id: int | None = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id", index=True)

    name: str = Field(max_length=255)
    partner_type: str | None = Field(default=None, max_length=100)
    contact_name: str | None = Field(default=None, max_length=255)
    contact_email: str | None = Field(default=None, max_length=255)
    notes: str | None = None

    # Extended fields for teaming board
    company_duns: str | None = Field(default=None, max_length=20)
    cage_code: str | None = Field(default=None, max_length=10)
    naics_codes: list = Field(default=[], sa_column=Column(JSON))
    set_asides: list = Field(default=[], sa_column=Column(JSON))
    capabilities: list = Field(default=[], sa_column=Column(JSON))
    clearance_level: str | None = Field(default=None, max_length=50)
    past_performance_summary: str | None = None
    website: str | None = Field(default=None, max_length=500)
    is_public: bool = Field(default=False)

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class RFPTeamingPartner(SQLModel, table=True):
    """
    Link table between RFPs and teaming partners.
    """

    __tablename__ = "rfp_teaming_partners"

    id: int | None = Field(default=None, primary_key=True)
    rfp_id: int = Field(foreign_key="rfps.id", index=True)
    partner_id: int = Field(foreign_key="teaming_partners.id", index=True)
    role: str | None = Field(default=None, max_length=255)

    created_at: datetime = Field(default_factory=datetime.utcnow)


class CaptureFieldType(str, Enum):
    TEXT = "text"
    NUMBER = "number"
    SELECT = "select"
    DATE = "date"
    BOOLEAN = "boolean"


class CaptureCustomField(SQLModel, table=True):
    """
    Custom fields for capture plans.
    """

    __tablename__ = "capture_custom_fields"

    id: int | None = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id", index=True)

    name: str = Field(max_length=255)
    field_type: CaptureFieldType = Field(default=CaptureFieldType.TEXT)
    options: list = Field(default=[], sa_column=Column(JSON))
    stage: CaptureStage | None = Field(default=None)
    is_required: bool = Field(default=False)

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class CaptureFieldValue(SQLModel, table=True):
    """
    Values for custom fields on a capture plan.
    """

    __tablename__ = "capture_field_values"

    id: int | None = Field(default=None, primary_key=True)
    capture_plan_id: int = Field(foreign_key="capture_plans.id", index=True)
    field_id: int = Field(foreign_key="capture_custom_fields.id", index=True)

    value: dict = Field(default={}, sa_column=Column(JSON))

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class CaptureCompetitor(SQLModel, table=True):
    """
    Competitive intelligence entries for an opportunity.
    """

    __tablename__ = "capture_competitors"

    id: int | None = Field(default=None, primary_key=True)
    rfp_id: int = Field(foreign_key="rfps.id", index=True)
    user_id: int = Field(foreign_key="users.id", index=True)

    name: str = Field(max_length=255)
    incumbent: bool = Field(default=False)
    strengths: str | None = None
    weaknesses: str | None = None
    notes: str | None = None

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class ActivityStatus(str, Enum):
    PLANNED = "planned"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    OVERDUE = "overdue"


class CaptureActivity(SQLModel, table=True):
    """
    Timeline activity for a capture plan (Gantt chart items).
    """

    __tablename__ = "capture_activities"

    id: int | None = Field(default=None, primary_key=True)
    capture_plan_id: int = Field(foreign_key="capture_plans.id", index=True)

    title: str = Field(max_length=500)
    start_date: date | None = None
    end_date: date | None = None
    is_milestone: bool = Field(default=False)
    status: ActivityStatus = Field(default=ActivityStatus.PLANNED)
    sort_order: int = Field(default=0)
    depends_on_id: int | None = Field(default=None, foreign_key="capture_activities.id")

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class TeamingRequestStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"


class TeamingRequest(SQLModel, table=True):
    """
    Teaming request between users and partners.
    """

    __tablename__ = "teaming_requests"

    id: int | None = Field(default=None, primary_key=True)
    from_user_id: int = Field(foreign_key="users.id", index=True)
    to_partner_id: int = Field(foreign_key="teaming_partners.id", index=True)
    rfp_id: int | None = Field(default=None, foreign_key="rfps.id", index=True)
    message: str | None = None
    status: TeamingRequestStatus = Field(default=TeamingRequestStatus.PENDING)

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class BidScorecardRecommendation(str, Enum):
    BID = "bid"
    NO_BID = "no_bid"
    CONDITIONAL = "conditional"


class ScorerType(str, Enum):
    AI = "ai"
    HUMAN = "human"


class BidScorecard(SQLModel, table=True):
    """Individual bid/no-bid scorecard from AI or a human team member."""

    __tablename__ = "bid_scorecards"

    id: int | None = Field(default=None, primary_key=True)
    rfp_id: int = Field(foreign_key="rfps.id", index=True)
    user_id: int = Field(foreign_key="users.id", index=True)

    criteria_scores: list[dict] = Field(default=[], sa_column=Column(JSON))
    overall_score: float | None = None
    recommendation: BidScorecardRecommendation | None = None
    confidence: float | None = None
    reasoning: str | None = Field(default=None, sa_column=Column(Text))

    scorer_type: ScorerType = Field(default=ScorerType.AI)
    scorer_id: int | None = None  # user_id if human scorer

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

File: frontend/src/types/capture.ts
import type { RFPStatus } from "./rfp";

// -----------------------------------------------------------------------------
// Capture Types
// -----------------------------------------------------------------------------

export type CaptureStage =
  | "identified"
  | "qualified"
  | "pursuit"
  | "proposal"
  | "submitted"
  | "won"
  | "lost";

export type BidDecision = "pending" | "bid" | "no_bid";

export type CaptureFieldType = "text" | "number" | "select" | "date" | "boolean";

export interface CapturePlan {
  id: number;
  rfp_id: number;
  owner_id: number;
  stage: CaptureStage;
  bid_decision: BidDecision;
  win_probability?: number | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface CaptureCustomField {
  id: number;
  name: string;
  field_type: CaptureFieldType;
  options: string[];
  stage?: CaptureStage | null;
  is_required: boolean;
  created_at: string;
  updated_at: string;
}

export interface CaptureFieldValue {
  field: CaptureCustomField;
  value?: unknown | null;
}

export interface CaptureFieldValueList {
  fields: CaptureFieldValue[];
}

export interface CapturePlanListItem extends CapturePlan {
  rfp_title: string;
  rfp_agency?: string | null;
  rfp_status?: RFPStatus | null;
}

export interface CaptureCompetitor {
  id: number;
  rfp_id: number;
  user_id: number;
  name: string;
  incumbent: boolean;
  strengths?: string | null;
  weaknesses?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface CaptureMatchInsight {
  plan_id: number;
  rfp_id: number;
  summary: string;
  factors: Array<{ factor: string; value: unknown }>;
}

export interface GateReview {
  id: number;
  rfp_id: number;
  reviewer_id: number;
  stage: CaptureStage;
  decision: BidDecision;
  notes?: string | null;
  created_at: string;
}

export interface TeamingPartner {
  id: number;
  name: string;
  partner_type?: string | null;
  contact_name?: string | null;
  contact_email?: string | null;
  notes?: string | null;
  company_duns?: string | null;
  cage_code?: string | null;
  naics_codes?: string[];
  set_asides?: string[];
  capabilities?: string[];
  clearance_level?: string | null;
  past_performance_summary?: string | null;
  website?: string | null;
  is_public?: boolean;
  created_at: string;
  updated_at: string;
}

export interface TeamingPartnerPublicProfile {
  id: number;
  name: string;
  partner_type?: string | null;
  contact_name?: string | null;
  contact_email?: string | null;
  company_duns?: string | null;
  cage_code?: string | null;
  naics_codes: string[];
  set_asides: string[];
  capabilities: string[];
  clearance_level?: string | null;
  past_performance_summary?: string | null;
  website?: string | null;
}

export type TeamingRequestStatus = "pending" | "accepted" | "declined";

export interface TeamingRequest {
  id: number;
  from_user_id: number;
  to_partner_id: number;
  rfp_id?: number | null;
  message?: string | null;
  status: TeamingRequestStatus;
  partner_name?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TeamingPartnerLink {
  id: number;
  rfp_id: number;
  partner_id: number;
  role?: string | null;
  created_at: string;
}

export type ActivityStatus = "planned" | "in_progress" | "completed" | "overdue";

export interface CaptureActivity {
  id: number;
  capture_plan_id: number;
  title: string;
  start_date?: string | null;
  end_date?: string | null;
  is_milestone: boolean;
  status: ActivityStatus;
  sort_order: number;
  depends_on_id?: number | null;
  created_at: string;
  updated_at: string;
}

export interface GanttPlanRow {
  plan_id: number;
  rfp_id: number;
  rfp_title: string;
  agency?: string | null;
  stage: CaptureStage;
  response_deadline?: string | null;
  activities: CaptureActivity[];
}

// -----------------------------------------------------------------------------
// Bid Scorecard Types
// -----------------------------------------------------------------------------

export type BidScorecardRecommendation = "bid" | "no_bid" | "conditional";
export type ScorerType = "ai" | "human";

export interface CriteriaScore {
  name: string;
  weight: number;
  score: number;
  reasoning?: string;
}

export interface BidScorecard {
  id: number;
  rfp_id: number;
  user_id: number;
  criteria_scores: CriteriaScore[];
  overall_score: number | null;
  recommendation: BidScorecardRecommendation | null;
  confidence: number | null;
  reasoning: string | null;
  scorer_type: ScorerType;
  scorer_id: number | null;
  created_at: string;
}

export interface BidDecisionSummary {
  rfp_id: number;
  total_votes: number;
  ai_score: number | null;
  human_avg: number | null;
  overall_recommendation: string | null;
  bid_count: number;
  no_bid_count: number;
  conditional_count: number;
}

// ---------------------------------------------------------------------------
// Capability Gap Analysis
// ---------------------------------------------------------------------------

export interface CapabilityGapItem {
  gap_type: string;
  description: string;
  required_value?: string | null;
  matching_partner_ids: number[];
}

export interface RecommendedPartner {
  partner_id: number;
  name: string;
  reason: string;
}

export interface CapabilityGapResult {
  rfp_id: number;
  gaps: CapabilityGapItem[];
  recommended_partners: RecommendedPartner[];
  analysis_summary: string;
}
